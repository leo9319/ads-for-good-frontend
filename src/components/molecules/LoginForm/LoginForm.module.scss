.form {
  display: inline-grid;
  gap: 12px;
  width: 100%;

  .forgotPass {
    line-height: var(--typography-line-height-3);
    font-size: var(--typography-font-size-3);
  }

  .loginBtn {
    margin-top: var(--spacing-sm);
    padding: var(--spacing-md) var(--spacing-lg) !important;
    font-size: var(--typography-font-size-4) !important;
    line-height: var(--typography-line-height-4) !important;
  }
}
