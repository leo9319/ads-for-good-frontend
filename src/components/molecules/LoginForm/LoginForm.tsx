import React, { ReactElement, useEffect, useState } from 'react';
import Link from '@radix-styles/atoms/Link';
import Flex from '@radix-styles/atoms/Flex';
import Button from '@radix-styles/atoms/Button';
import Input from '@components/atoms/Input';
import { FormErrorProps } from '@internal/types/common';
import { useForm, FieldValues } from 'react-hook-form';
import styles from './LoginForm.module.scss';
import CalloutMolecule, { CalloutProps } from '@radix-styles/molecules/Callout';

/**
 * @interface
 * The Forgot Password Form props type of {@link LoginForm}.
 */
interface ForgotPasswordObjProps {
  hidePassword?: boolean;
  emptyEmailWarning?: string;
}

/**
 * @interface
 * The Form props type of {@link LoginForm}.
 */
export interface LoginFormProps {
  /**
   * Placeholder for Email input for the login screen
   * @property
   */
  emailPlaceholder?: string;
  /**
   * Placeholder for Password input for the login screen
   * @property
   */
  passwordPlaceholder?: string;
  /**
   * Pass button text label, if need to display.
   * @property
   */
  text: string;
  /**
   * Text for the Hyperlink to forgot password screen
   * @property
   */
  forgotPassText?: string;
  /**
   * Function to handle form submission
   * @property
   */
  onSubmit: (data: FieldValues) => void;
  onClick?: () => void;
  /**
   * Errors that will be shown if any occur
   * @property
   */
  formError?: FormErrorProps;
  /**
   * It is to enable and disable the Spinner.
   * @property
   */
  loading?: boolean;
  /**
   * To Passing hyper link to navigate the page
   * @property
   */
  href?: string | undefined;
  /**
   * The object that helps to config the login form for Fotgot Password
   * @property
   */
  forgotPasswordObj?: ForgotPasswordObjProps;
  /**
   * The object that helps to show the server error message in callout
   * @property
   */
  calloutObj?: CalloutProps;
}

/**
 * LoginForm component
 *
 * @group Molecules
 * @category Component
 *
 * @param {LoginFormProps} props - The LoginForm props
 * @returns {ReactElement} - LoginForm Component
 */
export const LoginForm = ({
  text,
  forgotPassText,
  onSubmit,
  onClick,
  formError,
  loading = false,
  emailPlaceholder = '',
  passwordPlaceholder = '',
  href = '',
  forgotPasswordObj = {},
  calloutObj,
}: LoginFormProps): ReactElement => {
  const {
    register,
    handleSubmit,
    reset,
    getValues,
    setValue,
    setError,
    formState: { errors },
  } = useForm<FieldValues>({
    mode: 'onBlur',
  });
  const [status, setStatus] = useState<number | undefined>(undefined);

  const { hidePassword, emptyEmailWarning } = forgotPasswordObj;

  useEffect(() => {
    if (formError?.field && formError.message) {
      reset(getValues());
      setError(formError.field, {
        type: 'server',
        message: formError.message,
      });
      setStatus(formError.status);
    }
  }, [formError, reset, setError]);

  let updatedCalloutObj: CalloutProps = calloutObj || {
    type: 'error',
    isVisible: true,
    title: '',
    children: '',
  };

  if (
    errors.email?.type === 'required' &&
    hidePassword &&
    emptyEmailWarning &&
    updatedCalloutObj != null
  ) {
    updatedCalloutObj = {
      ...updatedCalloutObj,
      showIcon: true,
      type: 'error',
      isVisible: true,
      leftIcon: true,
      title: '',
      children: emptyEmailWarning,
    };
  }

  return (
    <form onSubmit={handleSubmit(onSubmit)} className={styles.form}>
      {updatedCalloutObj !== null && updatedCalloutObj?.children !== '' && (
        <CalloutMolecule {...updatedCalloutObj} />
      )}
      <Input
        label="Email Address"
        name="email"
        placeholder={emailPlaceholder}
        rules={{
          pattern: {
            message: 'Please check your email again for any errors.',
            value: /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/,
          },
          required: hidePassword
            ? 'This field is required'
            : 'Email is required',
        }}
        register={register}
        error={errors}
        status={status}
        type="text"
        setValue={setValue}
      />
      {!hidePassword && (
        <Input
          label="Password"
          name="password"
          placeholder={passwordPlaceholder}
          rules={{
            required: 'Password is required',
          }}
          register={register}
          error={errors}
          type="password"
          setValue={setValue}
        />
      )}
      {forgotPassText && !hidePassword && (
        <Flex display={'flex'} justify={'end'}>
          <Link
            onClick={onClick}
            targetHref={href}
            className={styles.forgotPass}
          >
            {forgotPassText}
          </Link>
        </Flex>
      )}
      <Button text={text} className={styles.loginBtn} disabled={loading} />
    </form>
  );
};

export default LoginForm;
