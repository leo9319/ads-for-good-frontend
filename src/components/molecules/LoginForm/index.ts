export type { LoginFormProps } from './LoginForm';
export { default } from './LoginForm';
export { LoginForm } from './LoginForm';
