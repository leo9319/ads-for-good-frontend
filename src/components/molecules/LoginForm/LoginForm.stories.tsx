import LoginForm from './LoginForm';
import { Meta, StoryObj } from '@storybook/react';

const meta = {
  title: 'Library/Molecules/LoginForm',
  component: LoginForm,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    emailPlaceholder: {
      control: { type: 'text' },
    },
    passwordPlaceholder: {
      control: { type: 'text' },
    },
    text: {
      control: { type: 'text' },
    },
    forgotPassText: {
      control: { type: 'text' },
    },
  },
  args: {
    forgotPassText: 'Forgot Your Password?',
    text: 'Login',
  },
} satisfies Meta<typeof LoginForm>;

export default meta;

type Story = StoryObj<typeof meta>;

export const SignIn: Story = {
  args: {
    emailPlaceholder: 'Email address',
    passwordPlaceholder: 'Password',
    forgotPassText: 'I forgot my password',
    onSubmit: data => console.log(data),
  },
};
