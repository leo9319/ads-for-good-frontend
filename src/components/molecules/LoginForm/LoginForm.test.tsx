import React from 'react';
import { render } from '@testing-library/react';
import LoginForm from './LoginForm';

describe('LoginForm', () => {
  it('Should render successfully', () => {
    const { baseElement } = render(
      <LoginForm
        forgotPassText="Forgot your password?"
        text="Sign In"
        onSubmit={data => console.log(data)}
      />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
