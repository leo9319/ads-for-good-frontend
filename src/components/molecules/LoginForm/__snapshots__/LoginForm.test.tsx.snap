// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Should render successfully 1`] = `
<body>
  <div>
    <form
      class="form"
    >
      <div
        class="formInput"
      >
        <label
          for="email"
        >
          Email Address
        </label>
        <input
          id="email"
          name="email"
          placeholder=""
          type="text"
        />
        <div
          class=""
        />
      </div>
      <div
        class="formInput showPassword"
      >
        <label
          for="password"
        >
          Password
        </label>
        <input
          id="password"
          name="password"
          placeholder=""
          type="password"
        />
        <div
          class="togglePassword"
        />
      </div>
      <div
        class="rt-Flex rt-r-display-flex rt-r-jc-end"
      >
        <a
          class="rt-Text rt-reset rt-Link rt-underline-auto link body-md forgotPass"
          data-accent-color=""
          tabindex="0"
        >
          Forgot your password?
        </a>
      </div>
      <button
        aria-label="Sign In"
        class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button wvButton xl primary none loginBtn"
        data-accent-color=""
      >
        <div
          class="buttonContent textOnly"
        >
          <span>
            Sign In
          </span>
        </div>
      </button>
    </form>
  </div>
</body>
`;
